torch/_functorch/partitioners.py

```diff
@@ -622,6 +622,74 @@ def insert_node_in_graph(node):
     return new_gm
 
 
+def apply_graphsafe_rng_functionalization(
+    fw_module: torch.fx.GraphModule,
+    bw_module: torch.fx.GraphModule,
+    fw_node: torch.fx.Node,
+    bw_node: torch.fx.Node,
+    device: torch.device,
+    rng_count: int,
+    last_fwd_input: torch.fx.Node,
+    last_bwd_input: torch.fx.Node,
+):
+    """Applies graph-safe RNG functionalization to forward and backward nodes in a CUDA device context.
+
+    This function modifies the forward and backward computation graphs by:
+    1. Creating RNG state placeholders for both passes
+    2. Updating the forward node to use graph-safe RNG state
+    3. Updating the backward node to use graph-safe RNG state
+    """
+    device_idx = device.index
+    assert device_idx is not None
+    fw_graph = fw_module.graph
+    bw_graph = bw_module.graph
+    graphsafe_run_with_rng_state = torch._prims.rng_prims.graphsafe_run_with_rng_state
+
+    # Handle forward pass
+    with fw_module.graph.inserting_after(last_fwd_input):
+        fwd_rng_state = fw_module.graph.placeholder(f"fwd_rng_state_{rng_count}")
+        fwd_rng_state.meta["val"] = torch.cuda.default_generators[
+            device_idx
+        ].graphsafe_get_state()
+        last_fwd_input = fwd_rng_state
+
+    # Handle backward pass
+    with bw_module.graph.inserting_after(last_bwd_input):
+        bwd_rng_state = bw_module.graph.placeholder(f"bwd_rng_state_{rng_count}")
+        bwd_rng_state.meta["val"] = torch.cuda.default_generators[
```

Comment:
Isn't this same as `fwd_rng_state.meta["val"]`? Just trying to understand. If yes, then perhaps CSE?

Reply:
sure, sounds good. although i think fake-ifying will fix our issues.